cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -536,6 +554,304 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+      std::for_each(thrust::make_counting_iterator(0),
+                    thrust::make_counting_iterator(pq_schema_elem.num_children),
+                    [&](auto const& idx) {
+                      co_walk_schemas(arrow_schema.children[idx], pq_schema_elem.children_idx[idx]);
+                    });
+
+      // true for DurationType columns only for now.
+      if (arrow_schema.type.id() != type_id::EMPTY) {
+        pq_schema_elem.arrow_type = arrow_schema.type.id();
+      }
+    };
+
+  // Get Parquet schema root
+  auto pq_schema_root = get_schema(0);
+
+  // verify equal number of children for both schemas at root level
+  if (pq_schema_root.num_children != static_cast<int32_t>(arrow_schema_root.children.size())) {
+    CUDF_LOG_ERROR("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // Verify equal number of children at all sub-levels
+  if (not std::all_of(thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(pq_schema_root.num_children),
+                      [&](auto const& idx) {
+                        return validate_schemas(arrow_schema_root.children[idx],
+                                                pq_schema_root.children_idx[idx]);
+                      })) {
+    CUDF_LOG_ERROR("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // All good, now co-walk schemas
+  std::for_each(thrust::make_counting_iterator(0),
+                thrust::make_counting_iterator(pq_schema_root.num_children),
+                [&](auto const& idx) {
+                  co_walk_schemas(arrow_schema_root.children[idx],
+                                  pq_schema_root.children_idx[idx]);
+                });
+}
+
+std::optional<std::string_view> aggregate_reader_metadata::decode_ipc_message(
+  std::string_view const serialized_message) const
+{
+  // Constants copied from arrow source and renamed to match the case
+  constexpr auto MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL         = sizeof(int32_t);
+  constexpr auto MESSAGE_DECODER_NEXT_REQUIRED_SIZE_METADATA_LENGTH = sizeof(int32_t);
+  constexpr int32_t IPC_CONTINUATION_TOKEN                          = -1;
+
+  // message buffer
+  auto message_buf = serialized_message.data();
+  // current message (buffer) size
+  auto message_size = static_cast<int32_t>(serialized_message.size());
+
+  // Check for empty message
+  if (message_size == 0) {
+    CUDF_LOG_ERROR("Parquet reader encountered zero length arrow:schema.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+  // Check for improper message.
+  if (message_size - MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL < 0) {
+    CUDF_LOG_ERROR("Parquet reader encountered unexpected arrow:schema message length.",
+                   "arrow:schema not processed.");
+  }
```

Comment:
should we return here? to me it looks like the memcpy below would read OOB if the message is too small
same goes for the similar code below

Reply:
Yes!

Reply:
All done!